apps/speed.c

```diff
@@ -908,79 +913,159 @@ static int EVP_Update_loop(void *args)
+                                     aead_ivlen, NULL)) {
+                BIO_printf(bio_err, "\nFailed to set iv length\n");
+                dofail();
+                exit(1);
+            }
         }
-    } else {
-        for (count = 0; COND(c[D_EVP][testnum]); count++) {
-            /* restore iv length field */
-            if (EVP_EncryptUpdate(ctx, NULL, &outl, NULL, lengths[testnum]) > 0
-                /* counter is reset on every update */
-                && EVP_EncryptUpdate(ctx, buf, &outl, buf, lengths[testnum]) > 0)
-                realcount++;
+        /* Set tag_len (Not for GCM/SIV at encryption stage) */
+        if (mode_op != EVP_CIPH_GCM_MODE
+            && mode_op != EVP_CIPH_SIV_MODE
+            && mode_op != EVP_CIPH_GCM_SIV_MODE) {
+            if (!EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_SET_TAG,
+                                     TAG_LEN, NULL)) {
+                BIO_printf(bio_err, "\nFailed to set tag length\n");
+                dofail();
+                exit(1);
+            }
+        }
+        if (!EVP_CipherInit_ex(ctx, NULL, NULL, key, aead_iv, -1)) {
+            BIO_printf(bio_err, "\nFailed to set key and iv\n");
+            dofail();
+            exit(1);
+        }
+        /* Set total length of input. Only required for CCM */
+        if (mode_op == EVP_CIPH_CCM_MODE) {
+            if (!EVP_EncryptUpdate(ctx, NULL, &outl,
+                                   NULL, lengths[testnum])) {
+                BIO_printf(bio_err, "\nCouldn't set input text length\n");
+                dofail();
+                exit(1);
+            }
         }
+        if (aead) {
+            if (!EVP_EncryptUpdate(ctx, NULL, &outl, aad, sizeof(aad))) {
+                BIO_printf(bio_err, "\nCouldn't insert AAD when encrypting\n");
+                dofail();
+                exit(1);
+            }
+        }
+        if (!EVP_EncryptUpdate(ctx, buf, &outl, buf, lengths[testnum])) {
+            BIO_printf(bio_err, "\nFailed to encrypt the data\n");
+            dofail();
+            exit(1);
+        }
+        if (EVP_EncryptFinal_ex(ctx, buf, &outl))
+            realcount++;
     }
-    if (decrypt)
-        final = EVP_DecryptFinal_ex(ctx, buf, &outl);
-    else
-        final = EVP_EncryptFinal_ex(ctx, buf, &outl);
-
-    if (final == 0)
-        BIO_printf(bio_err, "Error finalizing ccm loop\n");
     return realcount;
 }
 
 /*
  * To make AEAD benchmarking more relevant perform TLS-like operations,
  * 13-byte AAD followed by payload. But don't use TLS-formatted AAD, as
  * payload length is not actually limited by 16KB...
+ * CCM does not support streaming. For the purpose of performance measurement,
+ * each message is decrypted using the same (key,iv)-pair. Do not use this
+ * code in your application.
+ * For decryption, we will use buf2 to preserve the input text in buf.
  */
-static int EVP_Update_loop_aead(void *args)
+static int EVP_Update_loop_aead_dec(void *args)
```

Comment:
Could you please make it more readable, understandable and maintainable like following:

```
static int EVP_Update_loop_aead_dec(void *args)
{
    loopargs_t *loopArgs = *(loopargs_t **) args;
    unsigned char *inputBuf = loopArgs->buf;
    unsigned char *outputBuf = loopArgs->buf2;
    unsigned char *key = loopArgs->key;
    unsigned char tag[TAG_LEN];
    EVP_CIPHER_CTX *ctx = loopArgs->ctx;
    int outLen, count, realCount = 0;

    for (count = 0; COND(c[D_EVP][testnum]); count++) {
        if (!setIVLength(ctx)) return 0;
        if (!setTagLength(ctx)) return 0;
        if (!initializeCipher(ctx, key)) return 0;
        if (!setIV(ctx)) return 0;
        
        memcpy(tag, loopArgs->tag, TAG_LEN);

        if (!setTag(ctx, tag)) return 0;
        if (mode_op == EVP_CIPH_CCM_MODE && !setCipherTextLength(ctx)) return 0;
        if (aead && !insertAAD(ctx)) return 0;
        
        if (!decryptData(ctx, outputBuf, inputBuf)) return 0;
        if (EVP_DecryptFinal_ex(ctx, outputBuf, &outLen))
            realCount++;
    }
    return realCount;
}

static bool setIVLength(EVP_CIPHER_CTX *ctx) {
    if (mode_op != EVP_CIPH_SIV_MODE) {
        if (!EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_SET_IVLEN, aead_ivlen, NULL)) {
            BIO_printf(bio_err, "\nError: Failed to set IV length\n");
            dofail();
            return false;
        }
    }
    return true;
}

static bool setTagLength(EVP_CIPHER_CTX *ctx) {
    if (mode_op != EVP_CIPH_SIV_MODE && mode_op != EVP_CIPH_GCM_MODE && mode_op != EVP_CIPH_GCM_SIV_MODE) {
        if (!EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_SET_TAG, TAG_LEN, NULL)) {
            BIO_printf(bio_err, "\nError: Failed to set tag length\n");
            dofail();
            return false;
        }
    }
    return true;
}

static bool initializeCipher(EVP_CIPHER_CTX *ctx, const unsigned char *key) {
    if (!EVP_CipherInit_ex(ctx, NULL, NULL, key, aead_iv, -1)) {
        BIO_printf(bio_err, "\nError: Failed to set key and IV\n");
        dofail();
        return false;
    }
    return true;
}

static bool setIV(EVP_CIPHER_CTX *ctx) {
    if (mode_op != EVP_CIPH_SIV_MODE) {
        if (!EVP_DecryptInit_ex(ctx, NULL, NULL, NULL, aead_iv)) {
            BIO_printf(bio_err, "\nError: Failed to set IV\n");
            dofail();
            return false;
        }
    }
    return true;
}

static bool setTag(EVP_CIPHER_CTX *ctx, const unsigned char *tag) {
    if (!EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_SET_TAG, TAG_LEN, tag)) {
        BIO_printf(bio_err, "\nError: Failed to set tag\n");
        dofail();
        return false;
    }
    return true;
}

static bool setCipherTextLength(EVP_CIPHER_CTX *ctx) {
    if (!EVP_DecryptUpdate(ctx, NULL, &outLen, NULL, lengths[testnum])) {
        BIO_printf(bio_err, "\nError: Couldn't set ciphertext length\n");
        dofail();
        return false;
    }
    return true;
}

static bool insertAAD(EVP_CIPHER_CTX *ctx) {
    if (!EVP_DecryptUpdate(ctx, NULL, &outLen, aad, sizeof(aad))) {
        BIO_printf(bio_err, "\nError: Couldn't insert AAD during decryption\n");
        dofail();
        return false;
    }
    return true;
}

static bool decryptData(EVP_CIPHER_CTX *ctx, unsigned char *outputBuf, const unsigned char *inputBuf) {
    if (!EVP_DecryptUpdate(ctx, outputBuf, &outLen, inputBuf, lengths[testnum])) {
        BIO_printf(bio_err, "\nError: Failed to decrypt the data\n");
        dofail();
        return false;
    }
    return true;
}
```

This takes care of single responsibility principle -- Each function now handles a specific task, making the code easier to understand and maintain.

Reply:
This is a speed testing loop. Although in principle I agree with the "single responsibility principle", we should not take it fundamentally. I would not do this change. You could add more comments though to make the individual EVP API calls more clear. 

Also using thisNotationForFunctionNames would be wrong and not aligned with the OpenSSL code style.